rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+- Update references to dependent resources that remain in other modules to use data sources or [remote state](https://developer.hashicorp.com/terraform/language/state/remote-state-data)
+
+Note that production resources will remain in a single `prod` root module in the short term, but will likely eventually also be moved to team-based modules a later date. This will allow us to vet the changes proposed in this document, while also making sure we're adhering to the stricter security requirements for production that are discussed in the [Scalr workflow for production][scalr-workflow-production] section.
+
+### Root Module Naming Convention
+[root-module-naming-convention]: #root-module-naming-convention
+
+We'll also need to implement a new naming convention for root modules to keep them organized.
+
+Terraform's official [naming convention for published modules][terraform-module-publishing] is `terraform-<PROVIDER>-<NAME>`, where `<NAME>` can contain additional hyphens. Since our own root modules won't be published to Terraform's public registry, we can omit the `terraform-` prefix and use `<PROVIDER>-<NAME>` as a base convention, where `<NAME>` can contain any of the following:
+
+- AWS account nickname (e.g. `mbta-ctd-test`, `mbta-ctd-main`, or preferably a more succinct version that eliminates the `mbta-` prefix)
+- Team name (prefixed with `team-`)
+- Environment context (e.g. `dev`, `staging`, `prod`) - this overlaps with AWS account nickname, so we may want to discourage the use of both at the same time
+- Component descriptor of the type/category of resources the module is intended to contain, if it's one of multiple modules for the particular provider/account/context (e.g. `base`, `bootstrap`, `restricted`)
+- Nothing (in cases where one module encompasses all of a provider's configuration)
+
+A side-note regarding the AWS account nicknames: Our main AWS account is currently named "mbtacomsupport", but we are hoping to [rename it](https://docs.aws.amazon.com/accounts/latest/reference/manage-acct-alias.html) soon. We toyed with the name "mbta-ctd-prod", but since the account also contains development/staging resources&mdash;which can't reasonably be provisioned into their own separate "development/staging" AWS account&mdash;we've opted to recommend the name "mbta-ctd-main". That name is open to debate, however.
+
+#### Proposed Module Names
+
+With all the above in mind, the following are the proposed names for all existing and future root modules. (Note that the naming convention for child modules is different, and is covered in the [module versioning][guide-module-versioning] section.)
+
+Root modules associated with the "CTD Production" AWS account that should be renamed:
+
+- `bootstrap` → `aws-ctd-main-bootstrap`
+- `global` → `aws-ctd-main-base`
+- `dev` → multiple modules:
+  - move application development/staging resources to team-based modules:
```

Comment:
I touched on this briefly [here](https://github.com/mbta/technology-docs/blob/idw-terraform-workflow-improvements/rfcs/0000-terraform-workflow-improvements.md#alternatives-to-splitting-root-modules):

> One alternative approach would be to split modules by individual application instead of by team, but that would add complexity without adding much more in the way of flexibility.

For clarity, I'll expand that section to add that splitting by team allows us to manage Terraform (and Scalr) permissions at the team level, which I think is going to be a lot easier than splitting in other ways. I also worry that a team managing multiple smaller modules might be more granularity that we need at this point, though @bklebe is welcome to disagree with me on that. 😉 

Most of our application modules don't reference each other, and those that do could use [output values](https://developer.hashicorp.com/terraform/language/values/outputs), so there's no technical limitation here. There might be dependency issues, which we'll likely handle by prioritizing certain applications/teams higher in our disaster recovery plan.

Reply:
https://github.com/mbta/technology-docs/pull/16/commits/cccc1f12f453d93a52a68b6737e3373d312261b8